core/charm/repository/charmhub_test.go

```diff
@@ -1430,3 +1424,233 @@ func (s *composeSuggestionsSuite) setupMocks(c *gc.C) *gomock.Controller {
+
+	// Verify the selected base is stable (most stable risk).
+	c.Assert(result.origin.Platform.Channel, gc.Equals, "24.04/stable")
+
+	// Verify bases are sorted by stability: stable > candidate > beta > edge.
+	c.Assert(result.bases, gc.HasLen, 4)
+	c.Assert(result.bases[0].Channel, gc.Equals, "24.04/stable")
+	c.Assert(result.bases[1].Channel, gc.Equals, "24.04/candidate")
+	c.Assert(result.bases[2].Channel, gc.Equals, "24.04/beta")
+	c.Assert(result.bases[3].Channel, gc.Equals, "24.04/edge")
+}
+
+func (s *retryResolveWithRespBasesSuite) TestRetryResolveSortsBasesTrackThenRisk(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	repo := NewCharmHubRepository(s.logger, s.client)
+
+	// Setup error with mixed tracks and risks.
+	apiError := &transport.APIError{
+		Code:    transport.ErrorCodeInvalidCharmBase,
+		Message: "invalid charm base",
+		Extra: transport.APIErrorExtra{
+			DefaultBases: []transport.Base{
+				{Architecture: "amd64", Name: "ubuntu", Channel: "20.04/stable"},
+				{Architecture: "amd64", Name: "ubuntu", Channel: "24.04/beta"},
+				{Architecture: "amd64", Name: "ubuntu", Channel: "22.04/edge"},
+				{Architecture: "amd64", Name: "ubuntu", Channel: "24.04/stable"},
+				{Architecture: "amd64", Name: "ubuntu", Channel: "22.04/stable"},
+			},
+		},
+	}
+
+	// Expect refresh with 24.04/stable (highest track, most stable risk).
+	cfg, err := charmhub.InstallOneFromChannel("ubuntu-advantage", "stable", charmhub.RefreshBase{
+		Architecture: "amd64",
+		Name:         "ubuntu",
+		Channel:      "24.04/stable",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	s.client.EXPECT().Refresh(gomock.Any(), RefreshConfigMatcher{c: c, Config: cfg}).Return([]transport.RefreshResponse{{
+		InstanceKey: charmhub.ExtractConfigInstanceKey(cfg),
+		Entity: transport.RefreshEntity{
+			Type:         transport.CharmType,
+			ID:           "nt3kOo4yOEiUojrMTl09O9baTSBtTfsv",
+			Name:         "ubuntu-advantage",
+			Revision:     143,
+			MetadataYAML: "description: Ubuntu Advantage charm\nname: ubuntu-advantage\n",
+			ConfigYAML:   "options:\n  contract_url:\n",
+		},
+		EffectiveChannel: "stable",
+	}}, nil)
+
+	origin := corecharm.Origin{
+		Source: "charm-hub",
+		Platform: corecharm.Platform{
+			Architecture: "amd64",
+		},
+	}
+
+	result, err := repo.retryResolveWithRespBases("ubuntu-advantage", origin, apiError)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, gc.NotNil)
+
+	// Verify 24.04/stable is selected (highest track + most stable).
+	c.Assert(result.origin.Platform.Channel, gc.Equals, "24.04/stable")
+
+	// Verify sorting: 24.04/* > 22.04/* > 20.04/*, and within same track: stable > beta > edge.
+	c.Assert(result.bases, gc.HasLen, 5)
+	c.Assert(result.bases[0].Channel, gc.Equals, "24.04/stable")
+	c.Assert(result.bases[1].Channel, gc.Equals, "24.04/beta")
+	c.Assert(result.bases[2].Channel, gc.Equals, "22.04/stable")
+	c.Assert(result.bases[3].Channel, gc.Equals, "22.04/edge")
+	c.Assert(result.bases[4].Channel, gc.Equals, "20.04/stable")
```

Comment:
These should be checks not asserts (same with all this file). Asserts mean that we can't go any future, it must pass. An example of this, is if the length of the slice must be at least 5. The check method will continue and record an error. In this case, we could see just from the output of the test that the order was incorrect.

Reply:
will refactor this thanks, although I feel like we shouldn't be refactoring the asserts to checks for the rest of the file in this PR to avoid churn during the review, but perhaps I could do it for the rest of the file in a separate commit